### PR TITLE
Add utility functions for parsing certificates

### PR DESCRIPTION
**Pull request by xmudrii**
This PR is the first PR in the series of transition to the new logic for generating and regenerating etcdproxy certificates.

The full implementation can be found on the `certs-renewal` branch.

This PR adds two new functions for converting certificates from PEM to Certificates struct and back, as well as a function for validating the CA bundle.

### CODE REVIEW

File: pkg/certs/certs.go [hunk 1]
```diff
@@ -26,6 +30,65 @@ type Certificate struct {
...
+
+	if !okay {
+		return nil, fmt.Errorf("no valid certificate found")
+	}
+
+	return validCerts, nil
+}
+
+// GetPEMBytes converts x509-formatted Certificate and Key to the PEM bytes.
+// Key is allowed to be empty, as there is cases when we're not storing it (e.g. CA bundles).
```
Comment (deads2k) [inline pkg/certs/certs.go]: allowed to be empty because we may not have a key.

File: pkg/certs/certs.go [hunk 2]
```diff
@@ -26,6 +30,65 @@ type Certificate struct {
...
+		}
+		key = keyPairCert.PrivateKey
+	}
+
+	return &Certificate{certs, key}, nil
+}
+
+// ValidateCertificates checks are all certificates in the bundle valid, i.e. they have not expired.
+// The function returns new bundle with only valid certificates or error if no valid certificate is found.
+func ValidateCertificates(certs ...*x509.Certificate) ([]*x509.Certificate, error) {
```
Comment (deads2k) [inline pkg/certs/certs.go]: `FilterExpiredCerts`

File: pkg/certs/certs.go [hunk 3]
```diff
@@ -26,6 +30,65 @@ type Certificate struct {
...
+// ParseCertificateBytes converts PEM formatted certificate to the Certificate struct.
+// Key is allowed to be empty, as there is cases when we're not storing it (e.g. CA bundles).
+func ParseCertificateBytes(certBytes, keyBytes []byte) (*Certificate, error) {
+	certs, err := cert.ParseCertsPEM(certBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing certificates: %v", err)
+	}
+
+	var key crypto.PrivateKey
+	if keyBytes != nil {
```
Comment (deads2k) [inline pkg/certs/certs.go]: `len(keyBytes)`

**Comment by xmudrii** [comment]
Addressed all comments.

File: pkg/certs/certs_test.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+	if len(c.Certificates) != 1 {
+		t.Fatalf("expected 1 certificate in the chain, but got %d", len(c.Certificates))
+	}
+
+	validCerts, err := FilterExpiredCerts(c.Certificates...)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(validCerts) != 1 {
```
Comment (deads2k) [inline pkg/certs/certs_test.go]: you should have a negative test too. Shouldn't be too hard to get an expired cert.

**Comment by xmudrii** [comment]
Added.

File: pkg/certs/certs.go [hunk 5]
```diff
@@ -26,6 +30,65 @@ type Certificate struct {
...
+	var validCerts []*x509.Certificate
+	for _, c := range certs {
+		if c.NotAfter.After(currentTime) {
+			validCerts = append(validCerts, c)
+		}
+		okay = true
+	}
+
+	if !okay {
+		return nil, fmt.Errorf("no valid certificate found")
```
Comment (deads2k) [inline pkg/certs/certs.go]: This looks weird.  If you're filtering, why wouldn't you just return an empty slice?  That provides the same information to the caller and matches the method name.

File: pkg/certs/certs_test.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+		t.Fatalf("expected 1 valid certificate in the chain, but got %d", len(validCerts))
+	}
+}
+
+func TestValidateCertificatesExpired(t *testing.T) {
```
Comment (deads2k) [inline pkg/certs/certs_test.go]: make this test have both a valid and an invalid so you get back one.

**Comment by xmudrii** [comment]
Addressed all comments.

**Pull request merged**